File: packages/clickhouse/src/create-table-parser.ts
import {
  findMatchingParen,
  normalizeProjectionIndex,
  normalizeSQLFragment,
  splitTopLevelComma,
} from '@chkit/core'

type ProjectionDefinitionShape =
  | { name: string; query: string }
  | { name: string; index: string; type: string }

function parseClauseFromCreateTableQuery(
  createTableQuery: string | undefined,
  clausePattern: RegExp,
  stopPattern: RegExp
): string | undefined {
  if (!createTableQuery) return undefined
  // Table-level clauses (ENGINE, ORDER BY, PRIMARY KEY, ...) only appear after
  // the column list. Searching the whole query would match a keyword inside the
  // body — e.g. the `ORDER BY` of a projection's SELECT — and swallow the real
  // clause plus everything up to the next stop keyword (issue #190).
  const options = extractTableOptions(createTableQuery)
  const start = options.match(clausePattern)
  if (!start || start.index === undefined) return undefined
  const afterClause = options.slice(start.index + start[0].length)
  const stop = afterClause.match(stopPattern)
  const raw = (stop ? afterClause.slice(0, stop.index) : afterClause).trim()
  if (!raw) return undefined
  return normalizeSQLFragment(raw)
}

/**
 * Positions of the parens that open and close the column list — the close being
 * the one right before the table-level `ENGINE =`. Returns undefined when there
 * is no balanced column list (e.g. a view, or a query we can't parse).
 */
function findColumnListBounds(
  createTableQuery: string
): { open: number; close: number } | undefined {
  // Require a table-level `) ENGINE =` so we don't treat some other parenthesised
  // fragment as the column list.
  const engineMatch = /\)\s*ENGINE\s*=/i.exec(createTableQuery)
  if (!engineMatch || engineMatch.index === undefined) return undefined
  const openIndex = createTableQuery.indexOf('(')
  if (openIndex === -1) return undefined
  const close = findMatchingParen(createTableQuery, openIndex)
  return close === undefined ? undefined : { open: openIndex, close }
}

function extractCreateTableBody(createTableQuery: string | undefined): string | undefined {
  if (!createTableQuery) return undefined
  const bounds = findColumnListBounds(createTableQuery)
  if (!bounds) return undefined
  const body = createTableQuery.slice(bounds.open + 1, bounds.close).trim()
  return body.length > 0 ? body : undefined
}

/**
 * Everything after the column list: `ENGINE = ... PARTITION BY ... ORDER BY ...`.
 * This is where table-level clauses live. Falls back to the whole query when the
 * column list can't be located, preserving behaviour for unparseable inputs.
 */
function extractTableOptions(createTableQuery: string): string {
  const bounds = findColumnListBounds(createTableQuery)
  return bounds ? createTableQuery.slice(bounds.close + 1) : createTableQuery
}

export function parseSettingsFromCreateTableQuery(createTableQuery: string | undefined): Record<string, string> {
  if (!createTableQuery) return {}
  const settingsMatch = extractTableOptions(createTableQuery).match(/\bSETTINGS\b([\s\S]*?)(?:;|$)/i)
  if (!settingsMatch?.[1]) return {}
  const rawSettings = settingsMatch[1].trim()
  if (!rawSettings) return {}
  const items = splitTopLevelComma(rawSettings)
  const out: Record<string, string> = {}
  for (const item of items) {
    const eq = item.indexOf('=')
    if (eq === -1) continue
    const key = item.slice(0, eq).trim()
    const value = item.slice(eq + 1).trim()
    if (!key) continue
    out[key] = value
  }
  return out
}

export function parseTTLFromCreateTableQuery(createTableQuery: string | undefined): string | undefined {
  if (!createTableQuery) return undefined
  const ttlMatch = extractTableOptions(createTableQuery).match(/\bTTL\b([\s\S]*?)(?:\bSETTINGS\b|;|$)/i)
  const raw = ttlMatch?.[1]?.trim()
  if (!raw) return undefined
  return normalizeSQLFragment(raw)
}

export function parseEngineFromCreateTableQuery(createTableQuery: string | undefined): string | undefined {
  return parseClauseFromCreateTableQuery(
    createTableQuery,
    /\bENGINE\s*=\s*/i,
    /\bPRIMARY\s+KEY\b|\bORDER\s+BY\b|\bPARTITION\s+BY\b|\bUNIQUE\s+KEY\b|\bSAMPLE\s+BY\b|\bTTL\b|\bSETTINGS\b|;|$/i
  )
}

export function parsePrimaryKeyFromCreateTableQuery(
  createTableQuery: string | undefined
): string | undefined {
  return parseClauseFromCreateTableQuery(
    createTableQuery,
    /\bPRIMARY\s+KEY\b/i,
    /\bORDER\s+BY\b|\bPARTITION\s+BY\b|\bUNIQUE\s+KEY\b|\bSAMPLE\s+BY\b|\bTTL\b|\bSETTINGS\b|;|$/i
  )
}

export function parseOrderByFromCreateTableQuery(createTableQuery: string | undefined): string | undefined {
  return parseClauseFromCreateTableQuery(
    createTableQuery,
    /\bORDER\s+BY\b/i,
    /\bPRIMARY\s+KEY\b|\bPARTITION\s+BY\b|\bUNIQUE\s+KEY\b|\bSAMPLE\s+BY\b|\bTTL\b|\bSETTINGS\b|;|$/i
  )
}

export function parsePartitionByFromCreateTableQuery(
  createTableQuery: string | undefined
): string | undefined {
  return parseClauseFromCreateTableQuery(
    createTableQuery,
    /\bPARTITION\s+BY\b/i,
    /\bPRIMARY\s+KEY\b|\bORDER\s+BY\b|\bUNIQUE\s+KEY\b|\bSAMPLE\s+BY\b|\bTTL\b|\bSETTINGS\b|;|$/i
  )
}

export function parseUniqueKeyFromCreateTableQuery(
  createTableQuery: string | undefined
): string | undefined {
  return parseClauseFromCreateTableQuery(
    createTableQuery,
    /\bUNIQUE\s+KEY\b/i,
    /\bPRIMARY\s+KEY\b|\bORDER\s+BY\b|\bPARTITION\s+BY\b|\bSAMPLE\s+BY\b|\bTTL\b|\bSETTINGS\b|;|$/i
  )
}

export function parseProjectionsFromCreateTableQuery(
  createTableQuery: string | undefined
): ProjectionDefinitionShape[] {
  const body = extractCreateTableBody(createTableQuery)
  if (!body) return []
  const parts = splitTopLevelComma(body)
  const projections: ProjectionDefinitionShape[] = []
  for (const part of parts) {
    // Index-only projections have no SELECT body, so they must be matched
    // before the parenthesized SELECT form.
    const indexMatch = part.match(
      /^\s*PROJECTION\s+(?:`([^`]+)`|([A-Za-z_][A-Za-z0-9_]*))\s+INDEX\s+([\s\S]+?)\s+TYPE\s+([A-Za-z_][A-Za-z0-9_]*)\s*$/i
    )
    if (indexMatch) {
      const name = (indexMatch[1] ?? indexMatch[2] ?? '').trim()
      const index = normalizeProjectionIndex(indexMatch[3] ?? '')
      const type = (indexMatch[4] ?? '').trim()
      if (!name || !index || !type) continue
      projections.push({ name, index, type })
      continue
    }

    const match = part.match(
      /^\s*PROJECTION\s+(?:`([^`]+)`|([A-Za-z_][A-Za-z0-9_]*))\s*\(([\s\S]*)\)\s*$/i
    )
    if (!match) continue
    const name = (match[1] ?? match[2] ?? '').trim()
    const query = normalizeSQLFragment((match[3] ?? '').trim())
    if (!name || !query) continue
    projections.push({ name, query })
  }
  return projections
}

File: chkit_python/src/chkit/clickhouse/create_table_parser.py
"""Extract clauses from raw ``CREATE TABLE`` DDL strings.

1:1 port of ``packages/clickhouse/src/create-table-parser.ts``.

Used by ``listTableDetails`` (live ClickHouse introspection) and by
the future ``pull`` plugin to reconstruct a chkit schema from existing
tables. Pure parser — no I/O, no Pydantic models, no side effects.

The implementation matches the TS regex-based approach intentionally
rather than reaching for a full SQL parser: ClickHouse DDL has a
small, stable surface and an actual parser would carry far more
dependencies than the eight clauses we care about.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from chkit.core.key_clause import split_top_level_comma
from chkit.core.projection import normalize_projection_index
from chkit.core.sql_normalizer import normalize_sql_fragment

__all__ = [
    "ProjectionDefinitionShape",
    "parse_engine_from_create_table_query",
    "parse_order_by_from_create_table_query",
    "parse_partition_by_from_create_table_query",
    "parse_primary_key_from_create_table_query",
    "parse_projections_from_create_table_query",
    "parse_settings_from_create_table_query",
    "parse_ttl_from_create_table_query",
    "parse_unique_key_from_create_table_query",
]


@dataclass(frozen=True, slots=True)
class ProjectionDefinitionShape:
    """A SELECT projection (``query``) or an index-only projection (``index``/``type``)."""

    name: str
    query: str | None = None
    index: str | None = None
    type: str | None = None


_SETTINGS_RE = re.compile(r"\bSETTINGS\b(.*?)(?:;|$)", re.IGNORECASE | re.DOTALL)
_TTL_RE = re.compile(r"\bTTL\b(.*?)(?:\bSETTINGS\b|;|$)", re.IGNORECASE | re.DOTALL)
_BODY_ENGINE_RE = re.compile(r"\)\s*ENGINE\s*=", re.IGNORECASE)

_ENGINE_START = re.compile(r"\bENGINE\s*=\s*", re.IGNORECASE)
_ENGINE_STOP = re.compile(
    r"\bPRIMARY\s+KEY\b|\bORDER\s+BY\b|\bPARTITION\s+BY\b|\bUNIQUE\s+KEY\b"
    r"|\bSAMPLE\s+BY\b|\bTTL\b|\bSETTINGS\b|;|$",
    re.IGNORECASE,
)

_PRIMARY_KEY_START = re.compile(r"\bPRIMARY\s+KEY\b", re.IGNORECASE)
_PRIMARY_KEY_STOP = re.compile(
    r"\bORDER\s+BY\b|\bPARTITION\s+BY\b|\bUNIQUE\s+KEY\b|\bSAMPLE\s+BY\b"
    r"|\bTTL\b|\bSETTINGS\b|;|$",
    re.IGNORECASE,
)

_ORDER_BY_START = re.compile(r"\bORDER\s+BY\b", re.IGNORECASE)
_ORDER_BY_STOP = re.compile(
    r"\bPRIMARY\s+KEY\b|\bPARTITION\s+BY\b|\bUNIQUE\s+KEY\b|\bSAMPLE\s+BY\b"
    r"|\bTTL\b|\bSETTINGS\b|;|$",
    re.IGNORECASE,
)

_PARTITION_BY_START = re.compile(r"\bPARTITION\s+BY\b", re.IGNORECASE)
_PARTITION_BY_STOP = re.compile(
    r"\bPRIMARY\s+KEY\b|\bORDER\s+BY\b|\bUNIQUE\s+KEY\b|\bSAMPLE\s+BY\b"
    r"|\bTTL\b|\bSETTINGS\b|;|$",
    re.IGNORECASE,
)

_UNIQUE_KEY_START = re.compile(r"\bUNIQUE\s+KEY\b", re.IGNORECASE)
_UNIQUE_KEY_STOP = re.compile(
    r"\bPRIMARY\s+KEY\b|\bORDER\s+BY\b|\bPARTITION\s+BY\b|\bSAMPLE\s+BY\b"
    r"|\bTTL\b|\bSETTINGS\b|;|$",
    re.IGNORECASE,
)

_PROJECTION_RE = re.compile(
    r"^\s*PROJECTION\s+(?:`([^`]+)`|([A-Za-z_][A-Za-z0-9_]*))\s*\((.*)\)\s*$",
    re.IGNORECASE | re.DOTALL,
)

# Index-only projections have no SELECT body, so they must be matched before
# the parenthesized SELECT form.
_INDEX_PROJECTION_RE = re.compile(
    r"^\s*PROJECTION\s+(?:`([^`]+)`|([A-Za-z_][A-Za-z0-9_]*))"
    r"\s+INDEX\s+(.+?)\s+TYPE\s+([A-Za-z_][A-Za-z0-9_]*)\s*$",
    re.IGNORECASE | re.DOTALL,
)


def _parse_clause(
    query: str | None, start_pattern: re.Pattern[str], stop_pattern: re.Pattern[str]
) -> str | None:
    """Slice between ``start_pattern`` and the first ``stop_pattern`` hit."""
    if not query:
        return None
    # Table-level clauses (ENGINE, ORDER BY, PRIMARY KEY, ...) only appear
    # after the column list. Searching the whole query would match a keyword
    # inside the body — e.g. the `ORDER BY` of a projection's SELECT — and
    # swallow the real clause plus everything up to the next stop keyword
    # (issue #190).
    options = _extract_table_options(query)
    start = start_pattern.search(options)
    if start is None:
        return None
    after = options[start.end() :]
    stop = stop_pattern.search(after)
    raw = after[: stop.start()] if stop is not None else after
    raw = raw.strip()
    if not raw:
        return None
    return normalize_sql_fragment(raw)


def _find_column_list_bounds(query: str) -> tuple[int, int] | None:
    """Positions of the parens that open and close the column list.

    The close is the one right before the table-level ``ENGINE =``. Returns
    ``None`` when there is no balanced column list (e.g. a view, or a query we
    can't parse).
    """
    engine_match = _BODY_ENGINE_RE.search(query)
    if engine_match is None:
        return None
    # Up to and including the closing ')' before ENGINE.
    left = query[: engine_match.start() + 1]
    open_index = left.find("(")
    if open_index == -1:
        return None

    depth = 0
    in_string = False
    string_quote = "'"
    for i in range(open_index, len(left)):
        char = left[i]
        if not char:
            continue
        if in_string:
            if char == string_quote and (i == 0 or left[i - 1] != "\\"):
                in_string = False
            continue
        if char in {"'", '"'}:
            in_string = True
            string_quote = char
            continue
        if char == "(":
            depth += 1
            continue
        if char == ")":
            depth -= 1
            if depth == 0:
                return (open_index, i)
    return None


def _extract_create_table_body(query: str | None) -> str | None:
    """Return the body between the opening ``(`` and matching ``)`` before ENGINE."""
    if not query:
        return None
    bounds = _find_column_list_bounds(query)
    if bounds is None:
        return None
    body = query[bounds[0] + 1 : bounds[1]].strip()
    return body or None


def _extract_table_options(query: str) -> str:
    """Everything after the column list: ``ENGINE = ... ORDER BY ...``.

    This is where table-level clauses live. Falls back to the whole query when
    the column list can't be located, preserving behaviour for unparseable
    inputs.
    """
    bounds = _find_column_list_bounds(query)
    return query[bounds[1] + 1 :] if bounds is not None else query


def parse_settings_from_create_table_query(query: str | None) -> dict[str, str]:
    """Extract ``SETTINGS k=v, k=v`` as a dict (last write wins)."""
    if not query:
        return {}
    match = _SETTINGS_RE.search(_extract_table_options(query))
    if match is None:
        return {}
    raw = match.group(1).strip()
    if not raw:
        return {}
    out: dict[str, str] = {}
    for item in split_top_level_comma(raw):
        eq = item.find("=")
        if eq == -1:
            continue
        key = item[:eq].strip()
        value = item[eq + 1 :].strip()
        if not key:
            continue
        out[key] = value
    return out


def parse_ttl_from_create_table_query(query: str | None) -> str | None:
    if not query:
        return None
    match = _TTL_RE.search(_extract_table_options(query))
    if match is None:
        return None
    raw = match.group(1).strip()
    if not raw:
        return None
    return normalize_sql_fragment(raw)


def parse_engine_from_create_table_query(query: str | None) -> str | None:
    return _parse_clause(query, _ENGINE_START, _ENGINE_STOP)


def parse_primary_key_from_create_table_query(query: str | None) -> str | None:
    return _parse_clause(query, _PRIMARY_KEY_START, _PRIMARY_KEY_STOP)


def parse_order_by_from_create_table_query(query: str | None) -> str | None:
    return _parse_clause(query, _ORDER_BY_START, _ORDER_BY_STOP)


def parse_partition_by_from_create_table_query(query: str | None) -> str | None:
    return _parse_clause(query, _PARTITION_BY_START, _PARTITION_BY_STOP)


def parse_unique_key_from_create_table_query(query: str | None) -> str | None:
    return _parse_clause(query, _UNIQUE_KEY_START, _UNIQUE_KEY_STOP)


def parse_projections_from_create_table_query(
    query: str | None,
) -> list[ProjectionDefinitionShape]:
    body = _extract_create_table_body(query)
    if body is None:
        return []
    projections: list[ProjectionDefinitionShape] = []
    for part in split_top_level_comma(body):
        index_match = _INDEX_PROJECTION_RE.match(part)
        if index_match is not None:
            name = (index_match.group(1) or index_match.group(2) or "").strip()
            index = normalize_projection_index(index_match.group(3) or "")
            type_text = (index_match.group(4) or "").strip()
            if not name or not index or not type_text:
                continue
            projections.append(
                ProjectionDefinitionShape(name=name, index=index, type=type_text)
            )
            continue

        match = _PROJECTION_RE.match(part)
        if match is None:
            continue
        name = (match.group(1) or match.group(2) or "").strip()
        query_text = normalize_sql_fragment((match.group(3) or "").strip())
        if not name or not query_text:
            continue
        projections.append(ProjectionDefinitionShape(name=name, query=query_text))
    return projections
